crates/bevy_ecs/src/schedule/run_criteria.rs

```diff
@@ -0,0 +1,360 @@
+    }
+}
+
+impl<L> IntoRunCriteria<BoxedRunCriteriaLabel> for L
+where
+    L: RunCriteriaLabel,
+{
+    fn into(self) -> RunCriteriaDescriptorOrLabel {
+        RunCriteriaDescriptorOrLabel::Label(Box::new(self))
+    }
+}
+
+impl IntoRunCriteria<RunCriteria> for RunCriteria {
+    fn into(self) -> RunCriteriaDescriptorOrLabel {
+        RunCriteriaDescriptorOrLabel::Label(self.label)
+    }
+}
+
+pub trait RunCriteriaDescriptorCoercion {
+    /// Assigns a label to the criteria. Must be unique.
+    fn label(self, label: impl RunCriteriaLabel) -> RunCriteriaDescriptor;
+
+    /// Specifies that this criteria must be evaluated before a criteria with the given label.
+    fn before(self, label: impl RunCriteriaLabel) -> RunCriteriaDescriptor;
+
+    /// Specifies that this criteria must be evaluated after a criteria with the given label.
+    fn after(self, label: impl RunCriteriaLabel) -> RunCriteriaDescriptor;
+}
+
+impl RunCriteriaDescriptorCoercion for RunCriteriaDescriptor {
+    fn label(mut self, label: impl RunCriteriaLabel) -> RunCriteriaDescriptor {
+        self.label = Some(Box::new(label));
+        self
+    }
+
+    fn before(mut self, label: impl RunCriteriaLabel) -> RunCriteriaDescriptor {
+        self.before.push(Box::new(label));
+        self
+    }
+
+    fn after(mut self, label: impl RunCriteriaLabel) -> RunCriteriaDescriptor {
+        self.after.push(Box::new(label));
+        self
+    }
+}
+
+fn new_run_criteria_descriptor(system: BoxedSystem<(), ShouldRun>) -> RunCriteriaDescriptor {
+    RunCriteriaDescriptor {
+        system: RunCriteriaSystem::Single(system),
+        label: None,
+        before: vec![],
+        after: vec![],
+    }
+}
+
+impl RunCriteriaDescriptorCoercion for BoxedSystem<(), ShouldRun> {
+    fn label(self, label: impl RunCriteriaLabel) -> RunCriteriaDescriptor {
+        new_run_criteria_descriptor(self).label(label)
+    }
+
+    fn before(self, label: impl RunCriteriaLabel) -> RunCriteriaDescriptor {
+        new_run_criteria_descriptor(self).before(label)
+    }
+
+    fn after(self, label: impl RunCriteriaLabel) -> RunCriteriaDescriptor {
+        new_run_criteria_descriptor(self).after(label)
+    }
+}
+
+impl<S> RunCriteriaDescriptorCoercion for S
+where
+    S: System<In = (), Out = ShouldRun>,
+{
+    fn label(self, label: impl RunCriteriaLabel) -> RunCriteriaDescriptor {
+        new_run_criteria_descriptor(Box::new(self)).label(label)
+    }
+
+    fn before(self, label: impl RunCriteriaLabel) -> RunCriteriaDescriptor {
+        new_run_criteria_descriptor(Box::new(self)).before(label)
+    }
+
+    fn after(self, label: impl RunCriteriaLabel) -> RunCriteriaDescriptor {
+        new_run_criteria_descriptor(Box::new(self)).after(label)
+    }
+}
+
+pub struct RunCriteria {
+    label: BoxedRunCriteriaLabel,
+}
+
+impl RunCriteria {
+    /// Constructs a new run criteria that will retrieve the result of the criteria `label`
+    /// and pipe it as input to `system`.
```

Comment:
This needs documentation about the duplicate label problem if we're sticking with that behavior.

Reply:
Changed the behavior in 82db4f65e5eac318f2944852fd55cba88398c54f: it's now opt-in, and documented.